server/shared/src/main/scala/org/http4s/server/middleware/Throttle.scala

```diff
@@ -60,77 +61,77 @@ object Throttle {
-                        .map(if (_) Some(successResponse) else None)
-                    }
-
-                  attemptSet
+      val getNanoTime = G.monotonic.map(_.toNanos)
+      val refillEveryNanos = refillEvery.toNanos
+      for {
+        // make sure that refillEvery is positive
+        _ <- F.raiseUnless(refillEveryNanos > 0L)(
+          new IllegalArgumentException("refillEvery should be > 0 nano")
+        )
+        creationTime <- getNanoTime
+        counter <- F.ref((capacity.toDouble, creationTime))
+
+      } yield new TokenBucket[F] {
+        override def takeToken: F[TokenAvailability] =
+          for {
+            values <- counter.access
+            ((previousTokens, previousTime), setter) = values
+            currentTime <- getNanoTime
+            token <- {
+              val timeDifference = currentTime - previousTime
+              val tokensToAdd = timeDifference.toDouble / refillEveryNanos.toDouble
+              val newTokenTotal = Math.min(previousTokens + tokensToAdd, capacity.toDouble)
+
+              // If setter fails (yields the value false), then retry with recursive call
+              if (newTokenTotal >= 1)
+                setter((newTokenTotal - 1, currentTime))
+                  .ifM(F.pure(TokenAvailable: TokenAvailability), takeToken)
+              else {
+                def unavailable: TokenAvailability = {
+                  val timeToNextToken = refillEveryNanos - timeDifference
+                  TokenUnavailable(timeToNextToken.nanos.some)
                 }
-            }
-
-            def loop: F[TokenAvailability] =
-              attemptUpdate.flatMap { attempt =>
-                attempt.fold(loop)(token => token.pure[F])
+                setter((newTokenTotal, currentTime)).ifM(F.pure(unavailable), takeToken)
               }
-            loop
-          }
-        }
+            }
+          } yield token
       }
     }
   }
 
-  private def createBucket[F[_]](amount: Int, per: FiniteDuration)(implicit F: Temporal[F]) = {
+  private def createBucket[F[_]: Concurrent: Clock](amount: Int, per: FiniteDuration) = {
     val refillFrequency = per / amount.toLong
     TokenBucket.local(amount, refillFrequency)
   }
 
   /** Limits the supplied service to a given rate of calls using an in-memory [[TokenBucket]]
     *
     * @param amount the number of calls to the service to permit within the given time period.
-    * @param per the time period over which a given number of calls is permitted.
-    * @param http the service to transform.
+    * @param per    the time period over which a given number of calls is permitted.
+    * @param http   the service to transform.
     * @return a task containing the transformed service.
     */
-  def apply[F[_], G[_]](amount: Int, per: FiniteDuration)(
+  def apply[F[_]: Concurrent: Clock, G[_]](amount: Int, per: FiniteDuration)(
       http: Http[F, G]
-  )(implicit F: Temporal[F]): F[Http[F, G]] =
+  ): F[Http[F, G]] =
```

Comment:
Serializing my Discord question:
> if you have `Concurrent` and `Clock`, should you just reach for `Temporal`? `Temporal` is strictly stronger, but you could say the same when rewriting `Concurrent` and `Sync` to `Async`

Fabio responds:
> I would argue that it isn't _strictly_ stronger
> and therefore just switch to Temporal
> I can build a Scheduler with Concurrent and Clock 😛

Reply:
More seriously, I would stick to `Temporal` here. Upgrading from `Concurrent: Clock` to `Temporal` adds convenience and I'd argue doesn't significantly weaken parametricity. `Concurrent` already lets you add semantic blocking, so the fact that you have a new source of that with `Temporal` is not that significant.

Reply:
> `Concurrent` already lets you add semantic blocking

Because I had to think about it for a second: via waiting on a `Deferred`.

Reply:
Thanks both for chiming in. This seems to be one of those cases where the whole is more than the sum of its parts 🙂 I'll revert the constraints here

Reply:
How do you complete the Deferred without a dispatcher or spinning on the clock?

Reply:
Ok, to clarify. The first comment was really meant as a joke, because you can only build a Scheduler without Temporal by spinning. The second comment is serious though: the increase in power is from the ability to wait, it doesn't really matter imho if you're waiting because you're `sleep`ing or because you haven't received an http call yet. `Concurrent` lets you create `Deferred`s, and therefore gives you the ability to wait